eureka2-client/src/main/java/com/netflix/eureka2/client/interest/BatchAwareIndexRegistry.java

```diff
@@ -0,0 +1,123 @@
+                localBatchingRegistry.subscribe(notifications);
+
+                // Buffer is defined by merge of batch hint from the registry and the channel
+                final AtomicBoolean batchingMode = new AtomicBoolean();
+                Observable<ChangeNotification<T>> finishBatchingObservable = BatchFunctions.combine(
+                        localBatchingRegistry.forInterest(interest),
+                        remoteBatchingRegistry.forInterest(interest)
+                ).doOnNext(new Action1<Boolean>() {
+                    @Override
+                    public void call(Boolean status) {
+                        batchingMode.set(status);
+                    }
+                }).flatMap(new Func1<Boolean, Observable<ChangeNotification<T>>>() {
+                    @Override
+                    public Observable<ChangeNotification<T>> call(Boolean batching) {
+                        return batching ? Observable.<ChangeNotification<T>>empty() : Observable.just((ChangeNotification<T>) FINISH_BATCHING_NOTIFICATION);
+                    }
+                });
+
+                // Plain data change notification stream
+                Observable<ChangeNotification<T>> dataNotifications = notifications
+                        .filter(new Func1<ChangeNotification<T>, Boolean>() {
+                            @Override
+                            public Boolean call(ChangeNotification<T> notification) {
+                                return notification.isDataNotification();
+                            }
+                        }).flatMap(new Func1<ChangeNotification<T>, Observable<ChangeNotification<T>>>() {
+                            @Override
+                            public Observable<ChangeNotification<T>> call(ChangeNotification<T> notification) {
+                                ChangeNotification<T> result = notification;
+                                if (notification instanceof SourcedChangeNotification) {
+                                    result = ((SourcedChangeNotification<T>) notification).toBaseNotification();
+                                }
+                                if (notification instanceof SourcedModifyNotification) {
+                                    result = ((SourcedModifyNotification<T>) notification).toBaseNotification();
+                                }
+                                if (batchingMode.get()) {
+                                    return Observable.just(result);
+                                }
+                                return Observable.just(result, (ChangeNotification<T>) FINISH_BATCHING_NOTIFICATION);
+                            }
+                        });
+
+                PublishSubject<ChangeNotification<T>> resultSubject = PublishSubject.create();
+                resultSubject.subscribe(subscriber);
+                finishBatchingObservable.subscribe(resultSubject);
```

Comment:
You'll need to do finishBatchingObservable.mergeWith(dataNotifications).subscribe(resultSubject) here.